### ISSUE TITLE
[Question] Using appsettings.{environment}.json in Program.cs (Blazor WASM gRPC example)

### ISSUE THREAD

**Issue by JeepNL** [dotnet/aspnetcore#20442]
I'm using Blazor WebAssembly 3.2.0 Preview 3

 - Blazor .gRPC example: https://github.com/grpc/grpc-dotnet/tree/master/examples/Blazor
 - Microsoft Docs page: https://docs.microsoft.com/en-us/aspnet/core/blazor/dependency-injection?view=aspnetcore-3.1#blazor-webassembly

I'm now using `#if DEBUG #else #endif` for the string variable "`backendUrl`". I'd like to load that setting from appsettings.{environment}.json. I can get that after `var host = builder.Build();` (_info from Microsoft Docs, see sample code below, and link here above_) but the gRPC service is called before that.

My question: is it possible or should I keep using `#if DEBUG` etc. 
(_I'd like to use appsettings whenever it is possible anywhere in my code._)

**Part of my Program.cs**
```
	string backendUrl = string.Empty;
#if DEBUG
	backendUrl = "https://localhost:5001"; // Local debug URL
#else
	backendUrl = "https://<example>.com:5001"; // Production URL
#endif
	builder.Services.AddSingleton(services =>
	{
		// Create a gRPC-Web channel pointing to the backend server.
		// GrpcWebText is used because server streaming requires it. If server streaming is not used in your app
		// then GrpcWeb is recommended because it produces smaller messages.
		var httpClient = new HttpClient(new GrpcWebHandler(GrpcWebMode.GrpcWebText, new HttpClientHandler()));
		var channel = GrpcChannel.ForAddress(backendUrl, new GrpcChannelOptions { HttpClient = httpClient });
		return channel;
	});

	// load settings from appsettings.{environment}.json
	// see: https://docs.microsoft.com/en-us/aspnet/core/blazor/dependency-injection?view=aspnetcore-3.1#add-services-to-an-app
	var host = builder.Build();

	var backendDomain = host.Configuration["Settings:BackEndDomain"];
	Console.WriteLine($"Backend Domain: {backendDomain}");

	await host.RunAsync();

	// original
	// await builder.Build().RunAsync();
```

**Comment by javiercn** [comment]
@JeepNL thanks for contacting us.

We added support for config in preview3. You should be able to leverage it to configure your services in the latest preview. See https://devblogs.microsoft.com/aspnet/blazor-webassembly-3-2-0-preview-3-release-now-available/ for details

**Comment by JeepNL** [comment]
@javiercn Thank you for your reply. And yes, I'm using appsettings everywhere else in my Blazor Wasm 3.2.0 Preview 3 app (.Razor pages) . But my question is about "Propgram.cs". 

I can load the settings from appsettings in Program.cs, but my problem is that the gRPC service is called before that,

**Comment by JamesNK** [comment]
You should be able to get an `IConfiguration` inside the `AddSingleton`. e.g.

```cs
	builder.Services.AddSingleton(services =>
	{
        var configuration = services.GetRequiredService<IConfiguration>();
        var backendUrl = configuration["BackendUrl"];

		// Create a gRPC-Web channel pointing to the backend server.
		// GrpcWebText is used because server streaming requires it. If server streaming is not used in your app
		// then GrpcWeb is recommended because it produces smaller messages.
		var httpClient = new HttpClient(new GrpcWebHandler(GrpcWebMode.GrpcWebText, new HttpClientHandler()));
		var channel = GrpcChannel.ForAddress(backendUrl, new GrpcChannelOptions { HttpClient = httpClient });
		return channel;
	});
```

**Comment by JeepNL** [comment]
@JamesNK Thank you! I tried this and it works. I went to StackOverflow yesterday and posted my question there as well, as of now nobody was able to answer it. Normally I get a response in the first hour. I'll answer my own question there with a link to your answer here. 

Also, I now have a better understanding of how to load appsettings config in DI. I like to use appsetting.{environment}.json as much as I can because it makes switching between `'Development' & 'Production' easy. I'm coding with Visuals Studio on Windows and my production server is an Ubuntu 18.04 VPS. It compiles/runs on Ubuntu without making a single change in my code. I'm still smiling every single time when I do this.

And thank you for all those .NET gRPC samples at https://github.com/grpc/grpc-dotnet/tree/master/examples. I started with the gRPC Blazor example, and now I've added code from 'Ticketer' to 'protect' my (live) gRPC service. These examples are really helpful when you start with .NET gRPC. And because of their simplicity I learn a lot of other things as well. It's a great resource is what I'm saying 😉
